feat(slider): add custom range slider + SCSS improvements and layout fixes

DarkPix3l opened the pull request:
### Summary
This PR introduces a custom range slider component with dynamic fill and makes several style and performance-related improvements.

### Changes Included
- Added custom range slider with dynamic fill
- Refactored global SCSS to use `_helpers.scss` partial for better structure
-  Fixed layout latency by improving CSS and implementing flex/grid
-  Conditionally rendered title in ComponentFrame to avoid empty <h4>
-  Standardized layout using mixins for breakpoints and cleaned up layout styles

DarkPix3l committed:
refactor: move 'Autocomplete' component as standalone.

- Allow ComponentFrame to accept className for flexible grid placement (e.g., col-span, row-span, col-full) without breaking reusability

- Fix label display by setting label { display: block } to ensure proper stacking above inputs

- Enhance Autocomplete dropdown with max-height and overflow-y: scroll for long lists

DarkPix3l committed:
fix_style: rename global.scss to _helpers.scss to follow SCSS partial convention and clarify entry point structure

DarkPix3l committed:
fix: conditionally render title in ComponentFrame to avoid empty <h4>

DarkPix3l committed:
feat(slider): add custom range slider with dynamic fill

Initial draft of a custom range input component.
Uses internal state for value tracking, and calculates fill width dynamically.
Needs performance optimization.

DarkPix3l committed:
fix + add(slider): remove latency caused by CSS and implement flex-grow-based slider fill.
---
Removed transition-related lag from the fill bar and introduced a new approach using flex-grow, this approach can be reused also in a splitter component.

DarkPix3l committed:
Add  with standard mixins for breakpoints; layout tweaks

- Intended to improve test previews and make layout more consistent
- Applied additional layout/CSS tweaks for better visual structure
- Note:  breakpoint still needs refinement

DarkPix3l committed:
Stop tracking generated CSS files

DarkPix3l committed:
🧪  feat(slider): Add unit and integration tests

DarkPix3l committed:
⚙️ feat(Slider): remove internal/hybrid state, rely solely on external control

DarkPix3l committed:
feat(Slider): add keyboard arrow key support to increment/decrement value within bounds

DarkPix3l committed:
test(Slider): update tests to match controlled component behavior

DarkPix3l committed:
feat(Slider): add interactive and static stories showcasing different steps and boundary values.

pablo-maff commented on src/shared/components/ComponentFrame/Componentframe.tsx:
```diff
@@ -4,13 +4,15 @@ import style from './ComponentFrame.module.scss';
-      <h4 className={style.h4}>{title}</h4>
-      <div className={style.componentFrame}>{children}</div>
+    <div className={className} style={{ display: "flex", flexDirection: "column" }}>
+      {!!title?<h4 className={style.h4}>{title}</h4>: null}
```
eslint error, this works correctly but our current rules enforce a more idiomatic approach, like Boolean(title)

DarkPix3l replied:
```suggestion
      {title ? <h4 className={style.h4}>{title}</h4>: null}
```
Eslint still gives error(redundant boolean call) with `Boolean(title) ?`.  the `--fix` option has suggested this variant.
ESLint is now happy.

pablo-maff commented on src/shared/components/Slider/Slider.test.tsx:
```diff
@@ -0,0 +1,111 @@
+  };
+
+  beforeEach(() => {
+    mockOnChange.mockClear();
```
prefer to create a new spy on each call over re-using the same one after clearing its state. Clearing can be used in some scenario where we need to re-use the same spy with a clear state but it is something very rare

```suggestion
      let mockOnChange: ReturnType<typeof vi.fn>;

      beforeEach(() => {
        mockOnChange = vi.fn();
      });
```

pablo-maff commented on src/shared/components/Slider/Slider.test.tsx:
```diff
@@ -0,0 +1,111 @@
+
+  it('renders correctly and reflects initial value', () => {
+    render(<Slider {...defaultProps} />);
+    const slider = screen.getByRole('slider') as HTMLInputElement;
```
no need to use "as" here, TS already knows that slider is an HTMLInputElement. If you need a type assertion then you have a "code smell", it means that you are not coding defensively enough, e.g. not narrowing your types or that you are doing something wrong. IT should be avoided all the time unless there is a very good reason to use them.

DarkPix3l replied:
Done!

pablo-maff commented on src/shared/components/Slider/Slider.test.tsx:
```diff
@@ -0,0 +1,111 @@
+    expect(mockOnChange).toHaveBeenCalledWith(23);
+  });
+
+  it('respects min and max boundaries', () => {
```
It would be great if you could split this spec into two. Unit tests should only test one thing at the time, so having one that says "respects min boundary" and another one "respects max boundary" would be great

DarkPix3l replied:
done!

pablo-maff commented on src/shared/components/Slider/Slider.test.tsx:
```diff
@@ -0,0 +1,111 @@
+    expect(mockOnChange).toHaveBeenCalledWith(100);
+  });
+
+  it('handles keyboard arrow key presses correctly (controlled)', () => {
```
same than above, split it into specs that test only one case, having to use rerender is a symptom of this problem

DarkPix3l replied:
Splitting is fine, i'm on it.
Just one thought: here we repeat things like: `render(<Slider {...defaultProps} />)` and `getByRole('slider')` in multiple tests. Do you think it’s better to keep this duplication for test clarity, or should we extract it into a helper/setup function?

pablo-maff commented on src/shared/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,47 @@
+        className={style.slider_input}
+      />
+      <div className={style.track}>
+        <div className={style.filled} style={{ width: `${percentage}%` }} />
```
this is fine for a first iteration but I would like to improve it in the future to not use arbitrary inline styles across components. Probably something like a custom property or data attributes that could be read by scss can be the right direction to follow

DarkPix3l replied:
What do u think about a local type + css variable?

  ```js
   type StyleWithCustomProperty = React.CSSProperties & {
    '--filled-width': string;
  };

    const dynamicWidth: StyleWithCustomProperty = {
    '--filled-width': `${percentage}%`
  };
  ```

  ```css
  .filled {
  background: #53d3cc;
  width: var(--filled-width);
}
```

```html
<div className={style.filled} style={dynamicWidth} />

```

pablo-maff replied:
Excellent, this way if we look at the scss file without having context of the component we know straight away what's going on with the width

pablo-maff commented on src/shared/components/Slider/SliderFlex.tsx:
```diff
@@ -0,0 +1,52 @@
+  value?: number;
+}
+
+export default function Slider({
```
It has the same name than in Slider.tsx
Maybe this is a previous version of the slider that you forgot to delete?

DarkPix3l replied:
No, the export name is a mistake. 
I wanted to try a slider version using flex instead of calculation the width.
I kept this because i could use this method for another kind of component instead.
A splitter for example.

pablo-maff replied:
Ok, but here we should only have one. We could discuss on Saturday the pros and cons of both versions and decide which one to keep

pablo-maff commented on .gitignore:
```diff
@@ -25,4 +25,7 @@ dist-ssr
+coverage
+
+#Style - Temp build artifacts
+css
```
no need to ignore it here, with deleting the folder in your branch is enough

DarkPix3l replied:
Hey Pablo, thanks for the feedback!
Since these files are automatically transpiled from the Sass source whenever I make a change, they’re basically temporary. By ignoring them, I won’t have to worry about accidentally pushing them.
I think it’s cleaner to keep them out of the repo, and we can always regenerate them whenever needed.
Good call deleting them, thanks!

pablo-maff reviewed (changes requested):
Great work! 
We just need to polish some rough edges, besides the comments and review changes that i left there is also a couple of things to consider before pushing code to main from our feature or dev branch:

- We need to run prettier for consistent formatting, there is a command for it ```npm run format```. I recommend you to install it in your editor and enable the format on save option.
- ESLint should also be run before trying to merge to main. There is 1 error and 3 warnings now, we can be lenient while working on our dev branch but should be fixed before merging to main.

DarkPix3l committed:
fix: conditional rendering for title, according eslint standards

DarkPix3l committed:
Update src/shared/components/Slider/Slider.test.tsx

Co-authored-by: Pablo Maffioli <92764736+pablo-maff@users.noreply.github.com>

DarkPix3l committed:
Fix(slider test):

- Split the 'min/max boundaries' test into two dedicated specs.

- Split the 'keyboard' test into two separate specs to ensure each test case is fully isolated and to remove the need for `rerender` within a single test.

- Removed unnecessary type assertions (`as HTMLInputElement`) to improve type safety.

DarkPix3l reviewed (commented):
Thanks for the review. 
I've addressed the comments and pushed some changes.

DarkPix3l committed:
feat(slider): define custom CSS property type and move styles to SCSS

- Creat a custom type for the CSS variable to ensure TypeScript type safety and eliminate the need for type assertions.

- Removed inline styles for the <p> element, now handled via SCSS

DarkPix3l committed:
remove unused flex slider

DarkPix3l reviewed (commented):
Changes have been applied as requested, including deleting the 2nd slider. Ready for re-review.

pablo-maff commented:
> Changes have been applied as requested, including deleting the 2nd slider. Ready for re-review.

Thanks Greta, good job. Just a small detail missing, please run the format script when you can and we are good to go with this

DarkPix3l committed:
🧹 style: format codebase with Prettier & ESLint cleanup
